### PR TITLE
Document `nothing` type

### PR DESCRIPTION
**Pull request by allewun**
Addresses #673. 

This (https://github.com/facebook/hhvm/commit/5543eb2f765eb3fd988a899cc3797d30a8057b48 and https://github.com/facebook/hhvm/commit/5543eb2f765eb3fd988a899cc3797d30a8057b48) was the only official information I could find on `nothing`, please feel free to correct or amend with more details!

### CODE REVIEW

File: guides/hack/11-built-in-types/74-nothing.md [hunk 1]
```diff
@@ -0,0 +1,2 @@
+The `nothing` type is a bottom type and has the exact same semantics as the [`noreturn`](/hack/built-in-types/noreturn) keyword.
```
Comment (fredemmott) [inline guides/hack/11-built-in-types/74-nothing.md]: This isn't really accurate; if it has different typechecker behavior, it has different semantics.

Comment (allewun) [reply guides/hack/11-built-in-types/74-nothing.md]: True, though it's what I pulled out of the source. I don't really have any more context than those comments. Open to suggestions here - should I just remove that misleading clarification?

Comment (lexidor) [reply guides/hack/11-built-in-types/74-nothing.md]: I have only got my own experiences to rely on, but I can write something.

Comment (Wilfred) [reply guides/hack/11-built-in-types/74-nothing.md]: We plan to make `noreturn` the same as `nothing`. The nothing type is more general and is integrated into type inference. `noreturn` is an older implementation, isn't integrated into the type system, so has some limitations.

`noreturn` isn't a subtype of other types, so the following code only works for `nothing`:

```
function throw_error(): nothing {
  throw new Exception('oh dear');
}

function bool_to_int_or_error(bool $b): int {
  if ($b) {
    return 1;
  } else {
    return throw_error(); // this is an error for noreturn
  }
}
```

However, we have some dead code analysis for `noreturn` that isn't yet available for `nothing`, so the following code only works for `noreturn`:

```
function throw_error(): noreturn {
  throw new Exception('oh dear');
}

function bool_to_int_or_error(bool $b): int {
  if ($b) {
    return 1;
  } else {
    throw_error();
  }
}
```

Summary: users should be using `nothing` wherever possible, it's a proper bottom type.

**Comment by jjergus** [comment]
Closing this one since discussion has moved to #765

**Pull request closed without merging**